File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def _init_(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        d=defaultdict(list)
       
        def fun(root,d):
            if root==None:
                return
            if root.left!=None:
                d[root.val].append(root.left.val)
                
            if root.right!=None:
                d[root.val].append(root.right.val)
                
            fun(root.left,d)
            fun(root.right,d)
        fun(root,d)
        p=d.copy()
        print(p)
        for i in d.keys():
            a=d[i]
            for j in a:
                if i not in p[j]:
                    p[j].append(i)
        print(p)
        def dfs(rt,p,vis,c):
            for i in p[rt]:
                if vis[rt]!=0:
                    vis[rt]=1
                    dfs(i,p,vis,c+1)
            #vis[rt]=1
                    
        vis=[0]*(1000000+1)
        ans=0
        l=[[start,0]]
        c=0
        vis[start]=1
        while l:
            b=l[0]
            l.pop(0)
            c = max(c,b[1])
            for i in p[b[0]]:
                if not vis[i]:
                    vis[b[0]] = 1
                    l.append([i,b[1]+1])
                
                
        return c

File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    pair<int,bool> dfs(TreeNode*rt, int x){
        if(!rt)return {0,0};
        if(rt->val == x){
            auto [h1,v1] = dfs(rt->left,x);
            auto [h2,v2] = dfs(rt->right,x);
            ans = max({ans,h1,h2});
            return {1,1};
        }
        auto [h1,v1] = dfs(rt->left,x);
        auto [h2,v2] = dfs(rt->right,x);
        if(v1 || v2){
            int y = h1 + h2;
            ans = max(ans,y);
            if(v1)return {h1+1,1};
            else return {h2+1,1};
        }
        return {max(h1,h2)+1,0};
    }
    int amountOfTime(TreeNode* root, int start) {
        dfs(root,start);
        return ans;
    }
};